### ISSUE TITLE
Crash with slices e.g. `Annotated[int, 3:4]`

### ISSUE THREAD

**Issue by patrick-kidger** [python/mypy#10266]
**Bug Report / To Reproduce / Actual Behavior**

This:
```python
# file.py
from typing import Annotated
def func(x: Annotated[int, 3:4]):
    pass

# command line
mypy file.py
```
produces:
```
file.py:4: error: syntax error in type comment
Found 1 error in 1 file (errors prevented further checking)
```

**Your Environment**

- Mypy version used: 0.812
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9.2
- Operating system and version: Windows 10

**Further details**

Any use of slices seems to give the error, e.g. `Annotated[int, "a": 4]`.

I actually ran into this whilst implementing a type with a custom `__class_getitem__`, using slice notation as a convenient notation to pass pairs of values: `MyType["key1": value1, "key2": value2]`. Which I realise is a bit rogue -- I don't have any real hope of mypy integration for the type -- but it would at least be nice if it can be used without crashing mypy, with just a `# type: ignore` where it's imported.

**Comment by JelleZijlstra** [comment]
That's a fun one! It's probably pretty easy to fix if you track down where the error comes from (probably somewhere in `fastparse.py`).

**Comment by domdfcoding** [comment]
I did some digging. The error is being raised from `visit_Subscript` in `mypy.fastparse.TypeConverter`. The code is different depending on the Python version but the error is the same.

**On Python 3.9 and later:**

https://github.com/python/mypy/blob/2787d49633e732aa34de61c9ff429416ea10c857/mypy/fastparse.py#L1546-L1551

It is expecting the slice to be either:

* an `ast.Slice`
* an `ast.Tuple` which doesn't contain any `ast.Slice` nodes.

In this case, the slice is an `ast.Tuple` node containing:

* an `ast.Index` (the name `int`)
* an `ast.Slice` (`3:4`)

**On Python 3.8 and earlier:**

https://github.com/python/mypy/blob/2787d49633e732aa34de61c9ff429416ea10c857/mypy/fastparse.py#L1553-L1557

It is expecting the slice to be an `ast.Index` but instead it is an `ast.ExtSlice` containing:

* an `ast.Index` (the name `int`)
* an `ast.Slice` (`3:4`)

**Comment by amrit110** [comment]
Hi, Can I take this on? It's my very first issue! From what @domdfcoding has investigated, it seems that for py3.9, we can remove the constraint that the slice which is an `ast.Tuple` node not contain `ast.Slice` nodes. Looking at https://docs.python.org/3/library/ast.html#subscripting, I was wondering why that check was there in the first place since it says "It can be a Tuple and contain a Slice". For py3.8 and lower, the API documentation is not as clear as to what the slice can be but https://docs.python.org/3.8/library/ast.html#ast.NodeTransformer indicates that slice is probably `ast.Index`. So, for that I'm not sure, perhaps we can include that the slice can be `ast.ExtSlice` which contains `ast.Index`.

**Comment by PrasanthChettri** [comment]
@JelleZijlstra I am trying to fix the bug, for the example given above it's an instance of ast.Tuple(ast.Name, ast.Slice) and the problem is that of stated by @domdfcoding, and we will need to modify the conditions , it is probably not a good idea to just remove the constraint,  I will need a little bit of help with the new condition to replace it with.

**Comment by ap--** [comment]
It seems like fixing this issue in `mypy.fastparse.TypeConverter.visit_Subscript` would ultimately lead to allowing either both or at least one of:

```python
Dict[x:y]
Dict[int, x:y]
```

which is at least in one test explicitly marked to be a blocking error
 
https://github.com/python/mypy/blob/3dcccf221ead4b3143208fc2a1631336d6186e45/test-data/unit/check-fastparse.test#L320-L325

see: https://github.com/python/mypy/commit/cd5ff81f51f59976a31ca41f23e9cd70ece27f1c
and: https://github.com/python/mypy/pull/8716/commits/8ac365bc5a581bf1d148ee4ea7e84b3741e785ab

The docs for `typing.Annotated` https://docs.python.org/3/library/typing.html#typing.Annotated don't seem to suggest to use slicing syntax.
 
 
@patrick-kidger It looks as if the original use case is basically trying to emulate keyword arguments in getitem similar to pep472/pep637 via the slicing notation. So maybe https://www.python.org/dev/peps/pep-0472/#pass-a-dictionary-as-an-additional-index could be an easy solution to get this working in torchtyping with only a tiny overhead in characters:
 
 ```python
## the following does not throw an error ##
# file.py
from typing import Annotated
def func(x: Annotated[int, {3:4}]):  # pass a dict instead of slices
    pass

# command line
mypy file.py
```

Cheers,
Andreas 😃 

P.S.: If the decision is made to lift the currently in-place slice restriction, I can prepare a PR with a fix.

**Comment by morp007** [comment]
@patrick-kidger, there's no slice in your code. See python spec: [Slicings](https://docs.python.org/3/reference/expressions.html#slicings)

Your code should look like this:
```python
# file.py
from typing import Annotated
def func(x: Annotated[int, slice(3, 4)]):
    pass
```

**Comment by Zac-HD** [comment]
> ```python
> Dict[x:y]
> Dict[int, x:y]
> ```

IMO these should still be errors, just not (misleading) `syntax error in type comment` errors:

- "expects 2 type arguments, but 1 given" like `Dict[int]`), with a new "did you mean..." hint for slices.  
- "Invalid type comment or annotation" like `Dict[int, {1: 2}]`; which already "works" for custom types like `torchtyping.TensorType`.

and the tests could be updated accordingly, so that `mypy` continues to guide naive users towards the right way to express simple cases, without erroring out on more advanced or experimental uses.

**Comment by sobolevn** [comment]
I agree with @Zac-HD here. The error message should be improved, but `Dict[int, x:y]` and all of the example above should not be allowed (unless new PEPs or typing syntax chnages are made). I would be super confused to see `dict[x:y]` as a type.

Proposed altertives are:
1. `slice(1, 3)` instead of `1:3`
2. `{x: y}` instead of `x: y`

I will send a PR with the better error message today! 👍

**Comment by Zac-HD** [comment]
I don't think that #11241 should be considered as closing this issue without @sobolevn's proposed follow-up:

```python
from typing_extensions import Annotated

def func(x: Annotated[int, 3:4]):  # still a fatal error, contra (?) PEP-593
    x + "a string"                 # I expect an error here but don't get one
```

[As above, I think it's fine for slices to be an error in stdlib generics](https://github.com/python/mypy/issues/10266#issuecomment-929810067) - but IMO ought to be allowed in `Annotated` and third-party generic types.
